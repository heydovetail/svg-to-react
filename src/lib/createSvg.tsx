import * as React from "react";

export type CssValue = number | string;

export interface Size {
  width?: CssValue;
  height?: CssValue;
}

export interface Props {
  color?: string;
  size?: CssValue | Size;
}

/**
 * Build an SVG component
 *
 * Optimised for performance, both runtime execution and code minification. It
 * uses a named component (to enable DOM reconcilitation) and a render function
 * which uses positional arguments (for better minification).
 */
export function createSvg(
  render: (width: CssValue | undefined, height: CssValue | undefined, color: string) => React.ReactNode
) {
  return class SVG extends React.PureComponent<Props> {
    public render() {
      const { color = "currentColor", size = {} } = this.props;
      const width = typeof size === "object" ? size.width : size;
      const height = typeof size === "object" ? size.height : size;
      return <span style={{ lineHeight: 0, display: "block" }}>{render(width, height, color)}</span>;
    }
  };
}
