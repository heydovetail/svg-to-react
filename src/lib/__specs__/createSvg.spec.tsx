import "jest";
import * as React from "react";
import { createSvg } from "../createSvg";

describe(createSvg.name, () => {
  const SVG = createSvg((width, height, color) => <div style={{ width, height, color }} />);

  it("supports string size", () => {
    expect(<SVG size="100%" />).toMatchSnapshot();
  });

  it("supports number size", () => {
    expect(<SVG size={50} />).toMatchSnapshot();
  });
});
