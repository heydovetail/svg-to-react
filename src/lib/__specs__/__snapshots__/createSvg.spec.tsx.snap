// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSvg supports number size 1`] = `
<SVG
  size={50}
/>
`;

exports[`createSvg supports string size 1`] = `
<SVG
  size="100%"
/>
`;
